### ISSUE TITLE
use_self false positive on derive

### ISSUE THREAD

**Issue by lukasschlueter** [rust-lang/rust-clippy#3909]
Hi,

I think I've found a false positive for the use_self lint.
As far as I could tell, it's not covered by some other issue yet.

It appears that Clippy is running the lint for code that is getting `#derived`.
Here's an example:

```rust
#[macro_use]
extern crate relm_derive;

#[derive(Msg)]
enum M {
    Quit,
}
```

Output:

```bash
warning: unnecessary structure name repetition
 --> src/main.rs:5:6
  |
6 | enum M {
  |      ^ help: use the applicable keyword: `Self`
  |
  = note: requested on the command line with `-W clippy::use-self`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
```

```
$ cargo clippy -V
clippy 0.0.212 (92612c9 2019-03-18)
```

I'm not familiar with the way `#derive` works so I couldn't really reduce this down to a minimal example without dependencies.
However, I think the lint should not be applied to external dependencies.

Edit:
I tried to disable the lint for the enum with `#[allow(clippy::use_self)]` but it didn't have an effect:
```rust
#[derive(Msg)]
#[allow(clippy::use_self)]
enum M {
    Quit,
}
```

Kind regards,
Lukas
